### ISSUE TITLE
invalid-assignment false positive: dict literal narrowing fails through nested generic type alias in union

### ISSUE THREAD

**Issue by eitan-protego** [astral-sh/ty#3567]
### Summary

Dict literal key not narrowed to `Literal` when `Mapping[Literal[...], ...]` members are grouped behind an intermediate type alias

## Description

When multiple `Mapping[Literal[...], ...]` type aliases are combined in a **flat** union like `T | SelfOp[T] | ListOp[T]`, ty correctly narrows a dict literal's key type and accepts the assignment. However, if those same `Mapping` types are first grouped into an **intermediate type alias** (e.g. `type Ops[T] = SelfOp[T] | ListOp[T]`) and then used in the outer union (`T | Ops[T]`), ty no longer narrows the dict literal key and rejects the assignment with `invalid-assignment`.

This is related to #3487, whose fix in 0.0.39 covered the flat-union case but did not extend to nested aliases.

### Repro

```python
from typing import Literal
from collections.abc import Mapping, Sequence

type SelfOp[T] = Mapping[Literal["$eq", "$ne"], T]
type ListOp[T] = Mapping[Literal["$in", "$nin"], Sequence[T]]

# OK — flat union: dict literal key narrowed correctly
type FlatOp[T] = T | SelfOp[T] | ListOp[T]
a: FlatOp[str] = {"$in": ["a", "b"]}

# Error — same members grouped through an intermediate alias:
type Ops[T] = SelfOp[T] | ListOp[T]
type NestedOp[T] = T | Ops[T]
b: NestedOp[str] = {"$in": ["a", "b"]}
```

[ty playground](https://play.ty.dev/aa391ed9-ce7a-48f9-b0da-01c1e6f1d72a)

### Expected behavior

Both assignments should be accepted. `NestedOp[str]` flattens to the same type as `FlatOp[str]` (`str | Mapping[Literal["$eq", "$ne"], str] | Mapping[Literal["$in", "$nin"], Sequence[str]]`), so dict literal narrowing should work identically.

### Other type checkers

| Checker | `FlatOp` (line 9) | `NestedOp` (line 14) |
|---|---|---|
| pyright 1.1.400 | pass | pass |
| basedpyright 1.39.2 | pass | pass |
| pyrefly 0.64.1 | pass | pass |
| **mypy 1.20.2** | **error** | **error** |
| **ty 0.0.40** | pass | **error** |

Note: mypy rejects both forms (no bidirectional inference for dict literals through unions at all). pyright, basedpyright, and pyrefly accept both.

### Related

- #3487 — dict literal key not narrowed through union (fixed in 0.0.39 for flat unions; this issue covers the remaining nested-alias case)

### Version

ty 0.0.40 (7b95bc219 2026-05-27)
